### ISSUE TITLE
[Non-Feature Request] Disallow overloading of `!`

### ISSUE THREAD

**Issue by ChayimFriedman2** [wren-lang/wren#853]
Or provide a mechanism for custom truth.

The current behavior is inconsistent: you can override `!` and make your object "falsey", but you cannot really make it falsey because inside `if` statement it'll have a truthy value. There isn't a way to change that. So you end up with useless overloadable operator that just slow things down.

Instead, I suggest one of the following:

 - (Best IMHO) Make `!` a special operator that cannot be overloaded, and make a special opcode for it. It'll also speed up things a bit (although I agree that the `!` operator is, uhm, probably not your bottleneck).
 - Provide a way to make your object truthy, e.g. a method `isTruthy()`. This will:
   - Be a breaking change.
   - Slow down every `if`, `while` and `for` (yes, even `for`), which obviously _can_ be a bottleneck.
 - Use double `!!` to  evaluate the truthiness of an object. This will slow down things even more, but has the advantage of not being a breaking change (I think) and not force us to think about an appropriate name to the new method (naming is hard). Alternatively, we can just use one `!` and flip the result (`false` is falsey, everything else is truthy). This has almost the same performance characteristics as the previous approach, at the expense of losing some of the flexibility by requiring `!` to return `Bool`.

**Comment by ChayimFriedman2** [comment]
Yup. I know that, and specified the alternatives in the original post. In Python for example you overload `__bool__()`. But because of the performance implications, I don't think it's feasible. So instead to keep consistency I think we need to remove the `!` operator.

(And it won't require reentrancy, just like operators don't today).

**Comment by ChayimFriedman2** [comment]
> This also the reason why 0 is considered true in wren...

Hmm, no, this is not the reason. This is a design choice.

**Comment by mhermier** [comment]
What I meant was that it was not in the falsey list in the implementation
of the vm, not that it was a decision based on some form of logic.
